File: src/lexer.py
"""
Pseudocode Language - Lexer/Tokenizer
High-performance tokenization with minimal allocations
"""

from enum import IntEnum, auto
from dataclasses import dataclass
from typing import List, Optional
import re

class TokenType(IntEnum):
    # Literals
    INT = auto()
    FLOAT = auto()
    STRING = auto()
    TRUE = auto()
    FALSE = auto()
    
    # Identifiers & Keywords
    IDENT = auto()
    LET = auto()
    CONST = auto()
    FN = auto()
    RETURN = auto()
    IF = auto()
    THEN = auto()
    ELIF = auto()
    ELSE = auto()
    END = auto()
    WHILE = auto()
    FOR = auto()
    IN = auto()
    DO = auto()
    AND = auto()
    OR = auto()
    NOT = auto()
    
    # Operators
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    EQ = auto()
    NEQ = auto()
    LT = auto()
    GT = auto()
    LTE = auto()
    GTE = auto()
    ASSIGN = auto()
    ARROW = auto()
    RANGE = auto()
    
    # Bitwise
    BAND = auto()
    BOR = auto()
    BXOR = auto()
    SHL = auto()
    SHR = auto()
    
    # Delimiters
    LPAREN = auto()
    RPAREN = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    COMMA = auto()
    COLON = auto()
    NEWLINE = auto()
    
    # Special
    EOF = auto()
    ERROR = auto()

KEYWORDS = {
    'let': TokenType.LET,
    'const': TokenType.CONST,
    'fn': TokenType.FN,
    'return': TokenType.RETURN,
    'if': TokenType.IF,
    'then': TokenType.THEN,
    'elif': TokenType.ELIF,
    'else': TokenType.ELSE,
    'end': TokenType.END,
    'while': TokenType.WHILE,
    'for': TokenType.FOR,
    'in': TokenType.IN,
    'do': TokenType.DO,
    'and': TokenType.AND,
    'or': TokenType.OR,
    'not': TokenType.NOT,
    'true': TokenType.TRUE,
    'false': TokenType.FALSE,
}

@dataclass(slots=True)
class Token:
    type: TokenType
    value: str | int | float
    line: int
    col: int
    
    def __repr__(self):
        return f"Token({self.type.name}, {self.value!r}, {self.line}:{self.col})"

class Lexer:
    """High-performance lexer with single-pass tokenization"""
    
    __slots__ = ('source', 'pos', 'line', 'col', 'length')
    
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.col = 1
        self.length = len(source)
    
    def peek(self, offset: int = 0) -> str:
        """Look ahead without consuming"""
        idx = self.pos + offset
        return self.source[idx] if idx < self.length else '\0'
    
    def advance(self) -> str:
        """Consume and return current character"""
        ch = self.peek()
        self.pos += 1
        if ch == '\n':
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch
    
    def skip_whitespace(self):
        """Skip spaces and tabs (not newlines - they're significant)"""
        while self.peek() in ' \t\r':
            self.advance()
    
    def skip_comment(self):
        """Skip single-line and multi-line comments"""
        if self.peek() == '/' and self.peek(1) == '/':
            while self.peek() not in ('\n', '\0'):
                self.advance()
        elif self.peek() == '/' and self.peek(1) == '*':
            self.advance()  # /
            self.advance()  # *
            while not (self.peek() == '*' and self.peek(1) == '/'):
                if self.peek() == '\0':
                    return  # Unterminated comment
                self.advance()
            self.advance()  # *
            self.advance()  # /
    
    def make_token(self, type: TokenType, value, line: int, col: int) -> Token:
        return Token(type, value, line, col)
    
    def read_string(self) -> Token:
        """Parse string literal"""
        line, col = self.line, self.col
        quote = self.advance()  # Opening quote
        chars = []
        
        while self.peek() != quote and self.peek() != '\0':
            ch = self.advance()
            if ch == '\\':
                escape = self.advance()
                chars.append({
                    'n': '\n', 't': '\t', 'r': '\r',
                    '\\': '\\', '"': '"', "'": "'"
                }.get(escape, escape))
            else:
                chars.append(ch)
        
        if self.peek() == '\0':
            return self.make_token(TokenType.ERROR, "Unterminated string", line, col)
        
        self.advance()  # Closing quote
        return self.make_token(TokenType.STRING, ''.join(chars), line, col)
    
    def read_number(self) -> Token:
        """Parse integer or float literal"""
        line, col = self.line, self.col
        start = self.pos
        is_float = False
        
        # Handle hex/binary/octal
        if self.peek() == '0' and self.peek(1) in 'xXbBoO':
            self.advance()  # 0
            prefix = self.advance().lower()
            
            if prefix == 'x':
                while self.peek() in '0123456789abcdefABCDEF_':
                    self.advance()
                return self.make_token(TokenType.INT, int(self.source[start:self.pos].replace('_', ''), 16), line, col)
            elif prefix == 'b':
                while self.peek() in '01_':
                    self.advance()
                return self.make_token(TokenType.INT, int(self.source[start:self.pos].replace('_', ''), 2), line, col)
            elif prefix == 'o':
                while self.peek() in '01234567_':
                    self.advance()
                return self.make_token(TokenType.INT, int(self.source[start:self.pos].replace('_', ''), 8), line, col)
        
        # Decimal
        while self.peek().isdigit() or self.peek() == '_':
            self.advance()
        
        # Float part
        if self.peek() == '.' and self.peek(1).isdigit():
            is_float = True
            self.advance()  # .
            while self.peek().isdigit() or self.peek() == '_':
                self.advance()
        
        # Exponent
        if self.peek() in 'eE':
            is_float = True
            self.advance()
            if self.peek() in '+-':
                self.advance()
            while self.peek().isdigit():
                self.advance()
        
        num_str = self.source[start:self.pos].replace('_', '')
        if is_float:
            return self.make_token(TokenType.FLOAT, float(num_str), line, col)
        return self.make_token(TokenType.INT, int(num_str), line, col)
    
    def read_identifier(self) -> Token:
        """Parse identifier or keyword"""
        line, col = self.line, self.col
        start = self.pos
        
        while self.peek().isalnum() or self.peek() == '_':
            self.advance()
        
        text = self.source[start:self.pos]
        token_type = KEYWORDS.get(text, TokenType.IDENT)
        return self.make_token(token_type, text, line, col)
    
    def next_token(self) -> Token:
        """Get next token"""
        self.skip_whitespace()
        
        # Skip comments
        while self.peek() == '/' and self.peek(1) in '/*':
            self.skip_comment()
            self.skip_whitespace()
        
        line, col = self.line, self.col
        ch = self.peek()
        
        if ch == '\0':
            return self.make_token(TokenType.EOF, '', line, col)
        
        if ch == '\n':
            self.advance()
            return self.make_token(TokenType.NEWLINE, '\n', line, col)
        
        # String
        if ch in '"\'':
            return self.read_string()
        
        # Number
        if ch.isdigit():
            return self.read_number()
        
        # Identifier/keyword
        if ch.isalpha() or ch == '_':
            return self.read_identifier()
        
        # Two-character operators
        ch2 = self.peek(1)
        two_char = ch + ch2
        
        TWO_CHAR_OPS = {
            '==': TokenType.EQ,
            '!=': TokenType.NEQ,
            '<=': TokenType.LTE,
            '>=': TokenType.GTE,
            '->': TokenType.ARROW,
            '..': TokenType.RANGE,
            '<<': TokenType.SHL,
            '>>': TokenType.SHR,
        }
        
        if two_char in TWO_CHAR_OPS:
            self.advance()
            self.advance()
            return self.make_token(TWO_CHAR_OPS[two_char], two_char, line, col)
        
        # Single-character operators
        SINGLE_CHAR_OPS = {
            '+': TokenType.PLUS,
            '-': TokenType.MINUS,
            '*': TokenType.STAR,
            '/': TokenType.SLASH,
            '%': TokenType.PERCENT,
            '<': TokenType.LT,
            '>': TokenType.GT,
            '=': TokenType.ASSIGN,
            '&': TokenType.BAND,
            '|': TokenType.BOR,
            '^': TokenType.BXOR,
            '(': TokenType.LPAREN,
            ')': TokenType.RPAREN,
            '[': TokenType.LBRACKET,
            ']': TokenType.RBRACKET,
            ',': TokenType.COMMA,
            ':': TokenType.COLON,
        }
        
        if ch in SINGLE_CHAR_OPS:
            self.advance()
            return self.make_token(SINGLE_CHAR_OPS[ch], ch, line, col)
        
        # Unknown character
        self.advance()
        return self.make_token(TokenType.ERROR, f"Unexpected character: {ch}", line, col)
    
    def tokenize(self) -> List[Token]:
        """Tokenize entire source into list"""
        tokens = []
        while True:
            token = self.next_token()
            tokens.append(token)
            if token.type == TokenType.EOF:
                break
        return tokens


def tokenize(source: str) -> List[Token]:
    """Convenience function to tokenize source code"""
    return Lexer(source).tokenize()

File: wasm/src/compiler/lexer.ts
/*
 * Pseudocode WASM Compiler - Lexer
 * 
 * Tokenizes Pseudocode source code into a stream of tokens.
 * Ported from the C implementation in cvm/lexer.c
 */

/**
 * Token types for Pseudocode language.
 */
export enum TokenType {
    // Literals
    INT = 'INT',
    FLOAT = 'FLOAT',
    STRING = 'STRING',
    TRUE = 'TRUE',
    FALSE = 'FALSE',
    
    // Identifiers & Keywords
    IDENT = 'IDENT',
    LET = 'LET',
    CONST = 'CONST',
    FN = 'FN',
    RETURN = 'RETURN',
    IF = 'IF',
    THEN = 'THEN',
    ELIF = 'ELIF',
    ELSE = 'ELSE',
    END = 'END',
    WHILE = 'WHILE',
    FOR = 'FOR',
    IN = 'IN',
    TO = 'TO',
    DO = 'DO',
    AND = 'AND',
    OR = 'OR',
    NOT = 'NOT',
    MATCH = 'MATCH',
    CASE = 'CASE',
    TRY = 'TRY',
    CATCH = 'CATCH',
    FINALLY = 'FINALLY',
    THROW = 'THROW',
    CLASS = 'CLASS',
    EXTENDS = 'EXTENDS',
    SELF = 'SELF',
    SUPER = 'SUPER',
    NIL = 'NIL',
    ENUM = 'ENUM',
    
    // Advanced features
    YIELD = 'YIELD',
    ASYNC = 'ASYNC',
    AWAIT = 'AWAIT',
    STATIC = 'STATIC',
    FROM = 'FROM',
    AS = 'AS',
    MODULE = 'MODULE',
    EXPORT = 'EXPORT',
    IMPORT = 'IMPORT',
    
    // Operators
    PLUS = 'PLUS',
    MINUS = 'MINUS',
    STAR = 'STAR',
    SLASH = 'SLASH',
    PERCENT = 'PERCENT',
    EQ = 'EQ',
    NEQ = 'NEQ',
    LT = 'LT',
    GT = 'GT',
    LTE = 'LTE',
    GTE = 'GTE',
    ASSIGN = 'ASSIGN',
    ARROW = 'ARROW',
    RANGE = 'RANGE',
    
    // Delimiters
    LPAREN = 'LPAREN',
    RPAREN = 'RPAREN',
    LBRACKET = 'LBRACKET',
    RBRACKET = 'RBRACKET',
    LBRACE = 'LBRACE',
    RBRACE = 'RBRACE',
    COMMA = 'COMMA',
    DOT = 'DOT',
    COLON = 'COLON',
    UNDERSCORE = 'UNDERSCORE',
    NEWLINE = 'NEWLINE',
    
    // Special
    EOF = 'EOF',
    ERROR = 'ERROR',
}

/**
 * A token with its type, value, and location information.
 */
export interface Token {
    type: TokenType;
    value: string;
    line: number;
    column: number;
}

/**
 * Keywords map for quick lookup.
 */
const KEYWORDS: Map<string, TokenType> = new Map([
    ['let', TokenType.LET],
    ['const', TokenType.CONST],
    ['fn', TokenType.FN],
    ['return', TokenType.RETURN],
    ['if', TokenType.IF],
    ['then', TokenType.THEN],
    ['elif', TokenType.ELIF],
    ['else', TokenType.ELSE],
    ['end', TokenType.END],
    ['while', TokenType.WHILE],
    ['for', TokenType.FOR],
    ['in', TokenType.IN],
    ['to', TokenType.TO],
    ['do', TokenType.DO],
    ['and', TokenType.AND],
    ['or', TokenType.OR],
    ['not', TokenType.NOT],
    ['match', TokenType.MATCH],
    ['case', TokenType.CASE],
    ['try', TokenType.TRY],
    ['catch', TokenType.CATCH],
    ['finally', TokenType.FINALLY],
    ['throw', TokenType.THROW],
    ['class', TokenType.CLASS],
    ['extends', TokenType.EXTENDS],
    ['self', TokenType.SELF],
    ['super', TokenType.SUPER],
    ['nil', TokenType.NIL],
    ['true', TokenType.TRUE],
    ['false', TokenType.FALSE],
    ['enum', TokenType.ENUM],
    ['yield', TokenType.YIELD],
    ['async', TokenType.ASYNC],
    ['await', TokenType.AWAIT],
    ['static', TokenType.STATIC],
    ['from', TokenType.FROM],
    ['as', TokenType.AS],
    ['module', TokenType.MODULE],
    ['export', TokenType.EXPORT],
    ['import', TokenType.IMPORT],
]);

/**
 * Lexer for Pseudocode source code.
 */
export class Lexer {
    private source: string;
    private pos: number = 0;
    private line: number = 1;
    private column: number = 1;
    private tokens: Token[] = [];
    
    constructor(source: string) {
        this.source = source;
    }
    
    /**
     * Tokenize the entire source and return all tokens.
     */
    tokenize(): Token[] {
        this.tokens = [];
        
        while (!this.isAtEnd()) {
            this.skipWhitespaceAndComments();
            if (this.isAtEnd()) break;
            
            const token = this.scanToken();
            if (token) {
                this.tokens.push(token);
            }
        }
        
        this.tokens.push(this.makeToken(TokenType.EOF, ''));
        return this.tokens;
    }
    
    /**
     * Get the next token (for streaming use).
     */
    nextToken(): Token {
        this.skipWhitespaceAndComments();
        
        if (this.isAtEnd()) {
            return this.makeToken(TokenType.EOF, '');
        }
        
        return this.scanToken() || this.makeToken(TokenType.ERROR, this.advance());
    }
    
    private isAtEnd(): boolean {
        return this.pos >= this.source.length;
    }
    
    private peek(): string {
        if (this.isAtEnd()) return '\0';
        return this.source[this.pos];
    }
    
    private peekNext(): string {
        if (this.pos + 1 >= this.source.length) return '\0';
        return this.source[this.pos + 1];
    }
    
    private advance(): string {
        const char = this.source[this.pos++];
        if (char === '\n') {
            this.line++;
            this.column = 1;
        } else {
            this.column++;
        }
        return char;
    }
    
    private match(expected: string): boolean {
        if (this.isAtEnd() || this.source[this.pos] !== expected) {
            return false;
        }
        this.advance();
        return true;
    }
    
    private makeToken(type: TokenType, value: string): Token {
        return {
            type,
            value,
            line: this.line,
            column: this.column - value.length,
        };
    }
    
    private skipWhitespaceAndComments(): void {
        while (!this.isAtEnd()) {
            const char = this.peek();
            
            switch (char) {
                case ' ':
                case '\t':
                case '\r':
                    this.advance();
                    break;
                    
                case '\n':
                    // Newlines can be significant, emit NEWLINE token
                    // For now, just skip (the parser handles statement separation)
                    this.advance();
                    break;
                    
                case '/':
                    if (this.peekNext() === '/') {
                        // Single-line comment
                        while (!this.isAtEnd() && this.peek() !== '\n') {
                            this.advance();
                        }
                    } else {
                        return; // It's a division operator
                    }
                    break;
                    
                default:
                    return;
            }
        }
    }
    
    private scanToken(): Token | null {
        const startLine = this.line;
        const startColumn = this.column;
        const char = this.advance();
        
        // Single-character tokens
        switch (char) {
            case '(': return this.makeToken(TokenType.LPAREN, '(');
            case ')': return this.makeToken(TokenType.RPAREN, ')');
            case '[': return this.makeToken(TokenType.LBRACKET, '[');
            case ']': return this.makeToken(TokenType.RBRACKET, ']');
            case '{': return this.makeToken(TokenType.LBRACE, '{');
            case '}': return this.makeToken(TokenType.RBRACE, '}');
            case ',': return this.makeToken(TokenType.COMMA, ',');
            case ':': return this.makeToken(TokenType.COLON, ':');
            case '+': return this.makeToken(TokenType.PLUS, '+');
            case '*': return this.makeToken(TokenType.STAR, '*');
            case '%': return this.makeToken(TokenType.PERCENT, '%');
            case '_': 
                if (!this.isAlphaNumeric(this.peek())) {
                    return this.makeToken(TokenType.UNDERSCORE, '_');
                }
                // Fall through to identifier
                break;
        }
        
        // Two-character tokens
        switch (char) {
            case '-':
                if (this.match('>')) return this.makeToken(TokenType.ARROW, '->');
                return this.makeToken(TokenType.MINUS, '-');
                
            case '=':
                if (this.match('=')) return this.makeToken(TokenType.EQ, '==');
                return this.makeToken(TokenType.ASSIGN, '=');
                
            case '!':
                if (this.match('=')) return this.makeToken(TokenType.NEQ, '!=');
                return this.makeToken(TokenType.ERROR, '!');
                
            case '<':
                if (this.match('=')) return this.makeToken(TokenType.LTE, '<=');
                return this.makeToken(TokenType.LT, '<');
                
            case '>':
                if (this.match('=')) return this.makeToken(TokenType.GTE, '>=');
                return this.makeToken(TokenType.GT, '>');
                
            case '.':
                if (this.match('.')) return this.makeToken(TokenType.RANGE, '..');
                return this.makeToken(TokenType.DOT, '.');
                
            case '/':
                return this.makeToken(TokenType.SLASH, '/');
        }
        
        // String literals
        if (char === '"' || char === "'") {
            return this.scanString(char);
        }
        
        // Number literals
        if (this.isDigit(char)) {
            return this.scanNumber(char);
        }
        
        // Identifiers and keywords
        if (this.isAlpha(char) || char === '_') {
            return this.scanIdentifier(char);
        }
        
        return this.makeToken(TokenType.ERROR, char);
    }
    
    private scanString(quote: string): Token {
        const start = this.pos - 1;
        let value = '';
        
        while (!this.isAtEnd() && this.peek() !== quote) {
            if (this.peek() === '\\') {
                this.advance(); // Skip backslash
                if (!this.isAtEnd()) {
                    const escaped = this.advance();
                    switch (escaped) {
                        case 'n': value += '\n'; break;
                        case 't': value += '\t'; break;
                        case 'r': value += '\r'; break;
                        case '\\': value += '\\'; break;
                        case '"': value += '"'; break;
                        case "'": value += "'"; break;
                        default: value += escaped;
                    }
                }
            } else if (this.peek() === '\n') {
                return this.makeToken(TokenType.ERROR, 'Unterminated string');
            } else {
                value += this.advance();
            }
        }
        
        if (this.isAtEnd()) {
            return this.makeToken(TokenType.ERROR, 'Unterminated string');
        }
        
        this.advance(); // Closing quote
        return this.makeToken(TokenType.STRING, value);
    }
    
    private scanNumber(first: string): Token {
        let value = first;
        let isFloat = false;
        
        while (this.isDigit(this.peek())) {
            value += this.advance();
        }
        
        // Look for decimal part
        if (this.peek() === '.' && this.isDigit(this.peekNext())) {
            isFloat = true;
            value += this.advance(); // '.'
            while (this.isDigit(this.peek())) {
                value += this.advance();
            }
        }
        
        // Look for exponent
        if (this.peek() === 'e' || this.peek() === 'E') {
            isFloat = true;
            value += this.advance();
            if (this.peek() === '+' || this.peek() === '-') {
                value += this.advance();
            }
            while (this.isDigit(this.peek())) {
                value += this.advance();
            }
        }
        
        return this.makeToken(isFloat ? TokenType.FLOAT : TokenType.INT, value);
    }
    
    private scanIdentifier(first: string): Token {
        let value = first;
        
        while (this.isAlphaNumeric(this.peek())) {
            value += this.advance();
        }
        
        // Check for keyword
        const keyword = KEYWORDS.get(value.toLowerCase());
        if (keyword !== undefined) {
            return this.makeToken(keyword, value);
        }
        
        return this.makeToken(TokenType.IDENT, value);
    }
    
    private isDigit(char: string): boolean {
        return char >= '0' && char <= '9';
    }
    
    private isAlpha(char: string): boolean {
        return (char >= 'a' && char <= 'z') ||
               (char >= 'A' && char <= 'Z') ||
               char === '_';
    }
    
    private isAlphaNumeric(char: string): boolean {
        return this.isAlpha(char) || this.isDigit(char);
    }
}

/**
 * Tokenize source code and return tokens.
 */
export function tokenize(source: string): Token[] {
    const lexer = new Lexer(source);
    return lexer.tokenize();
}
